### PR TITLE
pci_devices: model and vendor information from system PCI db && add subsystem info

### PR DESCRIPTION
**Pull request by allanliu**
The current implementation linux's `pci_devices` table retrieves vendor and model information from sysFs attributes, which in turn utilizes systemd's hwdb to retrieve the information.  However hwdb utilizes a repo local copy of `pci.ids` to generate the `hwdb` rather than the system copy of `pci.ids` which usually far more updated.   This PR introduces retrieval of vendor and model information from the system copy of `pci.ids`.   New columns `subystem_vendor_id`, `subsystem_vendor`, `subsystem_model`, `subsystem_model_id` have also been added to enrich PCI device hardware information.

### CODE REVIEW

**Comment by theopolis** [comment]
ok to test

**Comment by guliashvili** [comment]
Hi @allanliu,

Thanks for your contribution.

You can resolve code audit error by running make format_master

**Comment by guliashvili** [comment]
@allanliu I think, it would help to change/format only the file changed by you. That way it's going to have a lot fewer conflicts.

**Comment by allanliu** [comment]
Updated.  Sorry about that.  Misunderstood your previous comment.

File: osquery/tables/system/linux/pci_devices.cpp [hunk 1]
```diff
@@ -26,38 +31,162 @@ const std::string kPCIKeyVendor = "ID_VENDOR_FROM_DATABASE";
...
 const std::string kPCIKeyID = "PCI_ID";
 const std::string kPCIKeyDriver = "DRIVER";
+const std::string kPCISubsysID = "PCI_SUBSYS_ID";
+
+const std::string kPciIdsPath = "/usr/share/misc/pci.ids";
+
+PciDB::PciDB(std::istream& db_filestream) {
+  // pci.ids keep track of subsystem information of vendor and models
+  // sequentially so we keep track of what the current vendor and models are.
+  std::string cur_vendor, cur_model, line;
```
Comment (guliashvili) [inline osquery/tables/system/linux/pci_devices.cpp]: It's better variables to have as small scope as possible. cur_vendor and cur_model can be declared in the while loop.

File: osquery/tables/system/linux/pci_devices.cpp [hunk 2]
```diff
@@ -26,38 +31,162 @@ const std::string kPCIKeyVendor = "ID_VENDOR_FROM_DATABASE";
...
+  while (std::getline(db_filestream, line)) {
+    if (line.size() < 7 || line.at(0) == '\n' || line.at(0) == '#') {
```
Comment (guliashvili) [inline osquery/tables/system/linux/pci_devices.cpp]: I would suggest
```
line = line.substr(0, line.find_first_of('#'))
boost:trim_right(line);
if(line.size() < 7){
   continue;
}
```
looks more readable and takes more corner cases into the consideration.

File: osquery/tables/system/linux/pci_devices.cpp [hunk 3]
```diff
@@ -26,38 +31,162 @@ const std::string kPCIKeyVendor = "ID_VENDOR_FROM_DATABASE";
...
+      cur_vendor = line.substr(0, 4);
+
+      // Once we get to illege vendor section we can stop since we're not
+      // currently parsing device classes.
+      if (cur_vendor == "ffff") {
+        return;
+      }
+
+      // Setup current vendor.
+      auto vendor = PciVendor();
```
Comment (guliashvili) [inline osquery/tables/system/linux/pci_devices.cpp]: I would suggest
auto& vendor = db_[cur_vendor]

File: osquery/tables/system/linux/pci_devices.cpp [hunk 4]
```diff
@@ -26,38 +31,162 @@ const std::string kPCIKeyVendor = "ID_VENDOR_FROM_DATABASE";
...
+      break;
+    }
+
+    case 1:
+      // Model info.
+      if (db_.find(cur_vendor) != db_.end() && line.size() > 7) {
+        cur_model = line.substr(1, 4);
+
+        // Set up current model under the current vendor.
+        auto model = PciModel();
```
Comment (guliashvili) [inline osquery/tables/system/linux/pci_devices.cpp]: auto& model = db_[cur_vendor].models[cur_model]

File: osquery/tables/system/linux/pci_devices.cpp [hunk 5]
```diff
@@ -26,38 +31,162 @@ const std::string kPCIKeyVendor = "ID_VENDOR_FROM_DATABASE";
...
+        model.id = cur_model;
+        // Bump 2 chars to account for whitespace separation.
+        model.desc = line.substr(7);
+        db_[cur_vendor].models[cur_model] = model;
```
Comment (guliashvili) [inline osquery/tables/system/linux/pci_devices.cpp]: Don't you need to check if cur_vendor is not empty first?

File: osquery/tables/system/linux/pci_devices.cpp [hunk 6]
```diff
@@ -26,38 +31,162 @@ const std::string kPCIKeyVendor = "ID_VENDOR_FROM_DATABASE";
...
+      } else {
+        VLOG(1) << "Unexpected error while parsing pci.ids: current vendor ID "
+                << cur_vendor << " does not exist in DB yet";
+      }
+
+      break;
+
+    case 2:
+      // Subsystem info;
+      if (db_.find(cur_vendor) != db_.end() &&
```
Comment (guliashvili) [inline osquery/tables/system/linux/pci_devices.cpp]: Don't access the more then once. Store iterator of the db_.find

File: osquery/tables/system/linux/pci_devices.cpp [hunk 7]
```diff
@@ -26,38 +31,162 @@ const std::string kPCIKeyVendor = "ID_VENDOR_FROM_DATABASE";
...
+          db_[cur_vendor].models.find(cur_model) !=
+              db_[cur_vendor].models.end() &&
+          line.size() > 11) {
+        // Store current subsystem information under current vendor and model.
+        db_[cur_vendor].models[cur_model].subsystemInfo[line.substr(2, 9)] =
+            line.substr(13);
```
Comment (guliashvili) [inline osquery/tables/system/linux/pci_devices.cpp]: It's not safe. You don't know that string is that large.

File: osquery/tables/system/linux/pci_devices.cpp [hunk 8]
```diff
@@ -26,38 +31,162 @@ const std::string kPCIKeyVendor = "ID_VENDOR_FROM_DATABASE";
...
+      break;
+
+    default:
+      VLOG(1) << "Unexpected pci.ids line format";
+    }
+  }
+}
+
+Status PciDB::getVendorName(const std::string& vendor_id, std::string& vendor) {
+  if (db_.find(vendor_id) == db_.end()) {
```
Comment (guliashvili) [inline osquery/tables/system/linux/pci_devices.cpp]: save the iterator and use it later.

File: osquery/tables/system/linux/pci_devices.cpp [hunk 9]
```diff
@@ -26,38 +31,162 @@ const std::string kPCIKeyVendor = "ID_VENDOR_FROM_DATABASE";
...
+  }
+
+  vendor = db_[vendor_id].name;
+  return Status(0, "OK");
+}
+
+Status PciDB::getModel(const std::string& vendor_id,
+                       const std::string& model_id,
+                       std::string& model) {
+  if (db_.find(vendor_id) == db_.end() ||
```
Comment (guliashvili) [inline osquery/tables/system/linux/pci_devices.cpp]: save the iterator and use it later.

File: osquery/tables/system/linux/pci_devices.cpp [hunk 10]
```diff
@@ -26,38 +31,162 @@ const std::string kPCIKeyVendor = "ID_VENDOR_FROM_DATABASE";
...
+
+    default:
+      VLOG(1) << "Unexpected pci.ids line format";
+    }
+  }
+}
+
+Status PciDB::getVendorName(const std::string& vendor_id, std::string& vendor) {
+  if (db_.find(vendor_id) == db_.end()) {
+    return Status(1, "Vendor ID does not exist");
```
Comment (guliashvili) [inline osquery/tables/system/linux/pci_devices.cpp]: Status::failure("Vendor ID does not exist");

File: osquery/tables/system/linux/pci_devices.cpp [hunk 11]
```diff
@@ -26,38 +31,162 @@ const std::string kPCIKeyVendor = "ID_VENDOR_FROM_DATABASE";
...
+  }
+
+  vendor = db_[vendor_id].name;
+  return Status(0, "OK");
```
Comment (guliashvili) [inline osquery/tables/system/linux/pci_devices.cpp]: Status::success()

File: osquery/tables/system/linux/pci_devices.cpp [hunk 12]
```diff
@@ -26,38 +31,162 @@ const std::string kPCIKeyVendor = "ID_VENDOR_FROM_DATABASE";
...
+}
+
+Status PciDB::getModel(const std::string& vendor_id,
+                       const std::string& model_id,
+                       std::string& model) {
+  if (db_.find(vendor_id) == db_.end() ||
+      db_[vendor_id].models.find(model_id) == db_[vendor_id].models.end()) {
+    return Status(1, "Vendor ID or Model ID does not exist");
```
Comment (guliashvili) [inline osquery/tables/system/linux/pci_devices.cpp]: Status::failure

File: osquery/tables/system/linux/pci_devices.cpp [hunk 13]
```diff
@@ -26,38 +31,162 @@ const std::string kPCIKeyVendor = "ID_VENDOR_FROM_DATABASE";
...
+  }
+
+  model = db_[vendor_id].models[model_id].desc;
+
+  return Status(0, "OK");
```
Comment (guliashvili) [inline osquery/tables/system/linux/pci_devices.cpp]: Status::success

File: osquery/tables/system/linux/pci_devices.cpp [hunk 14]
```diff
@@ -26,38 +31,162 @@ const std::string kPCIKeyVendor = "ID_VENDOR_FROM_DATABASE";
...
+Status PciDB::getSubsystemInfo(const std::string& vendor_id,
+                               const std::string& model_id,
+                               const std::string& subsystem_vendor_id,
+                               const std::string& subsystem_device_id,
+                               std::string& subsystem) {
+  auto subsystem_id = subsystem_vendor_id + " " + subsystem_device_id;
+
+  if (db_[vendor_id].models[model_id].subsystemInfo.find(subsystem_id) ==
+      db_[vendor_id].models[model_id].subsystemInfo.end()) {
+    return Status(
```
Comment (guliashvili) [inline osquery/tables/system/linux/pci_devices.cpp]: Status::failure

File: osquery/tables/system/linux/pci_devices.cpp [hunk 15]
```diff
@@ -26,38 +31,162 @@ const std::string kPCIKeyVendor = "ID_VENDOR_FROM_DATABASE";
...
+        1, "Subsystem ID does not exist in system pci.ids: " + subsystem_id);
+  }
+
+  subsystem = db_[vendor_id].models[model_id].subsystemInfo[subsystem_id];
+
+  return Status(0, "OK");
```
Comment (guliashvili) [inline osquery/tables/system/linux/pci_devices.cpp]: Status::success()

File: osquery/tables/system/linux/pci_devices.cpp [hunk 16]
```diff
@@ -26,38 +31,162 @@ const std::string kPCIKeyVendor = "ID_VENDOR_FROM_DATABASE";
...
+
+  vendor = db_[vendor_id].name;
+  return Status(0, "OK");
+}
+
+Status PciDB::getModel(const std::string& vendor_id,
+                       const std::string& model_id,
+                       std::string& model) {
+  if (db_.find(vendor_id) == db_.end() ||
+      db_[vendor_id].models.find(model_id) == db_[vendor_id].models.end()) {
```
Comment (guliashvili) [inline osquery/tables/system/linux/pci_devices.cpp]: save the iterator and use it later.

File: osquery/tables/system/linux/pci_devices.cpp [hunk 17]
```diff
@@ -26,38 +31,162 @@ const std::string kPCIKeyVendor = "ID_VENDOR_FROM_DATABASE";
...
+}
+
+Status PciDB::getSubsystemInfo(const std::string& vendor_id,
+                               const std::string& model_id,
+                               const std::string& subsystem_vendor_id,
+                               const std::string& subsystem_device_id,
+                               std::string& subsystem) {
+  auto subsystem_id = subsystem_vendor_id + " " + subsystem_device_id;
+
+  if (db_[vendor_id].models[model_id].subsystemInfo.find(subsystem_id) ==
```
Comment (guliashvili) [inline osquery/tables/system/linux/pci_devices.cpp]: There are no checks for key existence.

**Comment by guliashvili** [comment]
As you are parsing the input which can be modified by anyone, it's better to be very careful.

File: osquery/tables/system/linux/pci_devices.cpp (hunk 5 above)
Comment (allanliu) [reply osquery/tables/system/linux/pci_devices.cpp]: The check occurs at line `70`

File: osquery/tables/system/linux/pci_devices.cpp (hunk 17 above)
Comment (allanliu) [reply osquery/tables/system/linux/pci_devices.cpp]: Good catch. :+1:

File: osquery/tables/system/linux/pci_devices.cpp (hunk 7 above)
Comment (allanliu) [reply osquery/tables/system/linux/pci_devices.cpp]: Yeah, you're right, it's better to be safe than sorry.  I was only thinking about happy path and not potentially malformed pci.db being parsed.  I'll do a `line.substr(12)` with a `trim`

**Comment by guliashvili** [comment]
@allanliu  Thanks for the contribution and for reading my review. Will continue the review once those problems are fixed.

File: osquery/tables/system/linux/pci_devices.cpp (hunk 1 above)
Comment (allanliu) [reply osquery/tables/system/linux/pci_devices.cpp]: Logically, it does not work with `cur_vendor` and `cur_model` in the while loop.  B/c `pci.ids` reads top down, and the models and subsystem always belong to the their respective parents from the lines above, these variables are used to identify the vendor and model the current line belongs to (if vendor or model lines themselves)

**Comment by allanliu** [comment]
@guliashvili I think that should be all the changes requested (except for the issue commented above).  Please let me know if there's anything else.  Thanks!

File: osquery/tables/system/linux/pci_devices.cpp [hunk 18]
```diff
@@ -26,38 +31,191 @@ const std::string kPCIKeyVendor = "ID_VENDOR_FROM_DATABASE";
...
+      if (cur_vendor == "ffff") {
+        return;
+      }
+
+      // Setup current vendor.
+      auto& vendor = db_[cur_vendor];
+      vendor.id = cur_vendor;
+      // Bump 2 chars to account for whitespace separation..
+      vendor.name = line.substr(6);
+      db_[cur_vendor] = vendor;
```
Comment (guliashvili) [inline osquery/tables/system/linux/pci_devices.cpp]: It's not needed anymore.

File: osquery/tables/system/linux/pci_devices.cpp [hunk 19]
```diff
@@ -26,38 +31,191 @@ const std::string kPCIKeyVendor = "ID_VENDOR_FROM_DATABASE";
...
+    }
+
+    case 1: {
+      // Model info.
+      auto vendor = db_.find(cur_vendor);
+      if (vendor != db_.end() && line.size() > 7) {
+        cur_model = line.substr(1, 4);
+
+        // Set up current model under the current vendor.
+        auto& model = db_[cur_vendor].models[cur_model];
```
Comment (guliashvili) [inline osquery/tables/system/linux/pci_devices.cpp]: auto& model = vendor->second.models[cur_model] to save extra lookup time.

File: osquery/tables/system/linux/pci_devices.cpp [hunk 20]
```diff
@@ -26,38 +31,191 @@ const std::string kPCIKeyVendor = "ID_VENDOR_FROM_DATABASE";
...
+        model.id = cur_model;
+        // Bump 2 chars to account for whitespace separation.
+        model.desc = line.substr(7);
+        vendor->second.models[cur_model] = model;
```
Comment (guliashvili) [inline osquery/tables/system/linux/pci_devices.cpp]: not needed

File: osquery/tables/system/linux/pci_devices.cpp [hunk 21]
```diff
@@ -26,38 +31,191 @@ const std::string kPCIKeyVendor = "ID_VENDOR_FROM_DATABASE";
...
+    }
+
+    default:
+      VLOG(1) << "Unexpected pci.ids line format";
+    }
+  }
+}
+
+Status PciDB::getVendorName(const std::string& vendor_id, std::string& name) {
+  return vendor(vendor_id, [&](const PciVendor& v) {
```
Comment (guliashvili) [inline osquery/tables/system/linux/pci_devices.cpp]: I would kindly recommend getting rid of the vendor function(here and in other places). To me, It's okay to repeat two lines of code if it helps code readability.

File: osquery/tables/system/linux/pci_devices.cpp [hunk 22]
```diff
@@ -26,38 +31,191 @@ const std::string kPCIKeyVendor = "ID_VENDOR_FROM_DATABASE";
...
+  std::unique_ptr<udev_enumerate, decltype(del_udev_enum)> enumerate(
+      udev_enumerate_new(udev_handle.get()), del_udev_enum);
   if (enumerate.get() == nullptr) {
     VLOG(1) << "Could not get udev_enumerate handle";
     return results;
   }
 
+  std::ifstream raw(kPciIdsPath);
+  if (raw.fail()) {
+    LOG(ERROR) << "failed to read " << kPciIdsPath;
```
Comment (guliashvili) [inline osquery/tables/system/linux/pci_devices.cpp]: Could you add more information?  I like your format of the error message in other places mentioning pci.ids

File: osquery/tables/system/linux/pci_devices.cpp [hunk 23]
```diff
@@ -73,10 +231,57 @@ QueryData genPCIDevices(QueryContext& context) {
     // VENDOR:MODEL ID is in the form of HHHH:HHHH.
     std::vector<std::string> ids;
     auto device_id = UdevEventPublisher::getValue(device.get(), kPCIKeyID);
+
+    // pci.ids lower cases everything, so we follow suit.
+    std::transform(device_id.begin(),
```
Comment (guliashvili) [inline osquery/tables/system/linux/pci_devices.cpp]: boost::algorithm::to_lower

File: osquery/tables/system/linux/pci_devices.cpp [hunk 24]
```diff
@@ -73,10 +231,57 @@ QueryData genPCIDevices(QueryContext& context) {
...
+      if (pcidb.getModel(ids[0], ids[1], content).ok()) {
+        r["model"] = content;
+      }
+
+      // Try to enrich model with subsystem info.
+      std::vector<std::string> subsystem_ids;
+      auto subsystem_id =
+          UdevEventPublisher::getValue(device.get(), kPCISubsysID);
+
+      std::transform(subsystem_id.begin(),
```
Comment (guliashvili) [inline osquery/tables/system/linux/pci_devices.cpp]: boost::algorithm::to_lower

**Comment by guliashvili** [comment]
@allanliu Thank you for considering my review. I have just a few more comments and we should be good to go. 👍

File: osquery/tables/system/linux/pci_devices.cpp (hunk 18 above)
Comment (allanliu) [reply osquery/tables/system/linux/pci_devices.cpp]: Whops good catch.

**Comment by allanliu** [comment]
Updated.   Thanks for the rapid turnaround on the feedback :)

File: osquery/tables/system/linux/pci_devices.cpp [hunk 25]
```diff
@@ -26,38 +32,189 @@ const std::string kPCIKeyVendor = "ID_VENDOR_FROM_DATABASE";
 const std::string kPCIKeyModel = "ID_MODEL_FROM_DATABASE";
 const std::string kPCIKeyID = "PCI_ID";
 const std::string kPCIKeyDriver = "DRIVER";
+const std::string kPCISubsysID = "PCI_SUBSYS_ID";
+
+const std::string kPciIdsPath = "/usr/share/misc/pci.ids";
+
+PciDB::PciDB(std::istream& db_filestream) {
```
[the remark this answers is not in the document]
Comment (allanliu) [reply osquery/tables/system/linux/pci_devices.cpp]: Hmm...I like it.  The overall idea of it of makes sense to me (some parts I'm not sure fits or am misunderstanding).  I had thought that the constructor and each of it's case blocks was small enough that it's easy to reason about in just one function (which is why I didn't break up the cases), but if you had a hard time reading it, I obviously stand corrected :)
I will restructure it :+1:

**Comment by allanliu** [comment]
@guliashvili @akindyakov I restructured the constructor of `PciDB`.  Please let me know if it works.  Thanks!

File: osquery/tables/system/linux/pci_devices.cpp [hunk 26]
```diff
@@ -26,38 +32,245 @@ const std::string kPCIKeyVendor = "ID_VENDOR_FROM_DATABASE";
...
+  auto vendor_id = line.substr(0, 4);
+  // Setup current vendor.
+  // Bump 2 chars to account for whitespace separation..
+  auto result =
+      db_.emplace(vendor_id,
+                  PciVendor{vendor_id,
+                            line.substr(6),
+                            std::unordered_map<std::string, PciModel>{}});
+  if (result.second != true) {
+    return Status::failure("failed to save to db for line: " + line);
```
Comment (guliashvili) [inline osquery/tables/system/linux/pci_devices.cpp]: maybe add that key already exists.

File: osquery/tables/system/linux/pci_devices.cpp [hunk 27]
```diff
@@ -26,38 +32,245 @@ const std::string kPCIKeyVendor = "ID_VENDOR_FROM_DATABASE";
...
+  auto model_id = line.substr(1, 4);
+  // Set up current model under the current vendor.
+  // Bump 2 chars to account for whitespace separation.
+  auto result = cur_vendor->models.emplace(
+      model_id,
+      PciModel{model_id,
+               line.substr(7),
+               std::unordered_map<std::string, std::string>{}});
+  if (result.second != true) {
+    return Status::failure("failed to save to models db for line: " + line);
```
Comment (guliashvili) [inline osquery/tables/system/linux/pci_devices.cpp]: key already exists

File: osquery/tables/system/linux/pci_devices.cpp [hunk 28]
```diff
@@ -26,38 +32,245 @@ const std::string kPCIKeyVendor = "ID_VENDOR_FROM_DATABASE";
...
+const char kPciidsCommentChar = '#';
+
+Status PciDB::parseVendor(std::string& line, PciVendor*& cur_vendor) {
+  auto vendor_id = line.substr(0, 4);
+  // Setup current vendor.
+  // Bump 2 chars to account for whitespace separation..
+  auto result =
+      db_.emplace(vendor_id,
+                  PciVendor{vendor_id,
+                            line.substr(6),
```
Comment (guliashvili) [inline osquery/tables/system/linux/pci_devices.cpp]: scary substr

File: osquery/tables/system/linux/pci_devices.cpp [hunk 29]
```diff
@@ -26,38 +32,245 @@ const std::string kPCIKeyVendor = "ID_VENDOR_FROM_DATABASE";
...
 const std::string kPCIKeyDriver = "DRIVER";
+const std::string kPCISubsysID = "PCI_SUBSYS_ID";
+
+const std::string kPciidsPath = "/usr/share/misc/pci.ids";
+const std::string kPciidsDeviceClassStartIndicator = "ffff";
+const std::string kPciidsValidHexChars = "0123456789abcdef";
+const char kPciidsCommentChar = '#';
+
+Status PciDB::parseVendor(std::string& line, PciVendor*& cur_vendor) {
+  auto vendor_id = line.substr(0, 4);
```
[the remark this answers is not in the document]
Comment (allanliu) [reply osquery/tables/system/linux/pci_devices.cpp]: Yeah I agree.  I'll remove the length check from `parseLine` and put one in here.  It'll make the code easier to reason about too :)

**Comment by allanliu** [comment]
Moved the line check from `parseLine` to `parseVendor` and updated `Status::failure` messages..

**Pull request merged**